### ISSUE TITLE
Supports Cat's Xor as an Endpoint result

### ISSUE THREAD

**Issue by vkostyukov** [finagle/finch#504]
As [discussed](https://gitter.im/finagle/finch?at=5679ab54653b30761d75be97) on Gitter. It would be useful to support endpoints that return `Xor[A, B]` representing a failure and success case.

**Comment by ImLiar** [comment]
We've made the same thing with scalaz `\/` in glopart. I could port some code here.

But it would be nice to know how to work with left branch in the end, when we fold it it `Response`. I mean, should we just apply encoder to an error? And what about HTTP status codes?

**Comment by vkostyukov** [comment]
@ImLiar that's a great question actually. I probably need think this over, but for now I imagine this to be something like this:

```scala
case class Error(message: String)
val a: Endpoint[Xor[Error, Int]] = get("div" / int / int) { (a: Int, b: Int) =>
  if (b == 0) Output.payload(Xor.left(Error("Can't divide by 0")), Status.BadRequest)
  else Ok(a / b)
}
```

So basically this should work fine (not need to change `io.finch.internal.*`) as long as we have `EncodeResponse` defined for `Xor[A: EncodeResponse, B: EncodeResponse]`.

**Comment by ImLiar** [comment]
Still need a folder for `Xor[Response, Response] => Response`, but yea, I got it.

**Comment by ImLiar** [comment]
Well, it would be nice to pick `Content-Type` from instance of `EncodeResponse` which is applied to `A` or `B`, depending on `Xor` content. But with current API it seems impossible:

```
new EncodeResponse[Xor[A, B]] {

      override def apply(rep: Xor[A, B]): Buf = ???
      override def contentType: String = ???

    }
```

**Comment by vkostyukov** [comment]
Fixed in #518.
